src/extension/baser.ts

```diff
@@ -0,0 +1,171 @@
+        }
+        return this.validatedPathsLocalToArtifact.get(localPath) ?? localPath;
+    }
+
+    // Hacky: We are using `openTextDocument` to test the existence of documents as VS Code does not provide a dedicated existence API.
+    // The similar Node `fs` API does not resolve custom URI schemes in the same way that VS Code does otherwise we would use that.
+    private async pathExists(path: string) {
+        try {
+            await workspace.openTextDocument(Uri.parse(path));
+        } catch (error) {
+            return false;
+        }
+        return true;
+    }
+
+    private activeInfoMessages = new Set<string>() // Prevent repeat message animations when arrowing through many results with the same uri.
+    public async translateArtifactToLocal(artifactPath: string) { // Retval is validated.
+        // Temp.
+        if (artifactPath.startsWith('sarif:')) return artifactPath;
+        const validateUri = async () => {
+            // Cache
+            if (this.validatedPathsArtifactToLocal.has(artifactPath))
+                return this.validatedPathsArtifactToLocal.get(artifactPath);
+
+            // File System Exist
+            if (await this.pathExists(artifactPath))
+                return artifactPath;
+
+            // Known Bases
+            for (const [artifactBase, localBase] of this.basesArtifactToLocal) {
+                if (!artifactPath.startsWith(artifactBase)) continue; // Just let it fall through?
+
+                const normalizedArtifactPath = `${artifactPath.startsWith('/') ? '' : '/'}${artifactPath}`;
+                const localPath = normalizedArtifactPath.replace(artifactBase, localBase);
+
+                if (await this.pathExists(localPath)) {
+                    this.updateValidatedPaths(artifactPath, localPath);
+                    return localPath;
+                }
+            }
+
+            { // API-injected baseUris
+                const localPath = await this.tryUriBases(artifactPath);
+                if (localPath) return localPath;
+            }
+
+            // Distinct Project Items
+            const {file} = artifactPath;
+            if (this.distinctLocalNames.has(file) && this.store.distinctArtifactNames.has(file)) {
+                const localPath = this.distinctLocalNames.get(file)!; // Not undefined due to surrounding if.
+                this.updateValidatedPaths(artifactPath, localPath);
+                this.updateBases(artifactPath.split('/'), localPath.split('/'));
+                return localPath;
+            }
+
+            // Open Docs
+            for (const doc of workspace.textDocuments) {
+                const localPath = doc.uri.path;
+                if (localPath.file !== artifactPath.file) continue;
+                this.updateValidatedPaths(artifactPath, localPath);
+                this.updateBases(artifactPath.split('/'), localPath.split('/'));
+                return localPath;
+            }
+
+            return ''; // Can't find uri.
```

Comment:
>'' [](start = 19, length = 2)

Just curious as to why '' vs undefined?

Seems like undefined would force callers to handle the case where the string cannot be found without accidentally forgetting to check for empty string.

Reply:
Both approaches could work. Ideally I should lock the return type to just `string` (and not `undefined`). This way later `if(!validatedUri)` checks below become explicit no-empty-string checks.

---
In reply to: [441634671](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441634671) [](ancestors = 441634671)

Reply:
Constrained the typing to just `string`.

---
In reply to: [442427277](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r442427277) [](ancestors = 442427277,441634671)